src/main/java/com/google/devtools/build/lib/vfs/Blake3MessageDigest.java

```diff
@@ -0,0 +1,229 @@
+package com.google.devtools.build.lib.vfs;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.hash.Funnel;
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hasher;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.security.DigestException;
+import java.security.MessageDigest;
+
+public final class Blake3MessageDigest extends MessageDigest implements Hasher {
+  // These constants match the native definitions in:
+  // https://github.com/BLAKE3-team/BLAKE3/blob/master/c/blake3.h
+  public static final int KEY_LEN = 32;
+  public static final int OUT_LEN = 32;
+
+  // To reduce the number of calls made via JNI, buffer up to this many bytes.
+  // If a call to "hash()" is made and less than this much data has been
+  // written, a single JNI call will be made that initializes, hashes, and
+  // cleans up the hasher, rather than making separate calls for each operation.
+  public static final int ONESHOT_THRESHOLD = 8 * 1024;
+  private static ThreadLocal<ByteBuffer> threadLocalBuffer = new ThreadLocal<ByteBuffer>();
```

Comment:
Why use thread local? I think it's possible to have two instances of this class running on the same thread, or the same instance could be used across threads.

I think creating a buffer per instance should be good enough.

Reply:
Sure, fine for now. Can revisit performance more later.